cluster/cluster.go

```diff
@@ -0,0 +1,472 @@
+// THE SOFTWARE.
+
+package cluster
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/m3db/m3em/node"
+
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/shard"
+	"github.com/m3db/m3x/errors"
+	"github.com/m3db/m3x/log"
+)
+
+var (
+	errInsufficientCapacity          = fmt.Errorf("insufficient node capacity in environment")
+	errNodeAlreadyUsed               = fmt.Errorf("unable to add node, already in use")
+	errNodeNotInUse                  = fmt.Errorf("unable to remove node, not in use")
+	errClusterNotUnitialized         = fmt.Errorf("unable to setup cluster, it is not unitialized")
+	errClusterUnableToInitialize     = fmt.Errorf("unable to initialize cluster, it needs to be setup")
+	errClusterUnableToAlterPlacement = fmt.Errorf("unable to alter cluster placement, it needs to be setup/running")
+	errUnableToStartUnsetupCluster   = fmt.Errorf("unable to start cluster, it has not been setup")
+	errClusterUnableToTeardown       = fmt.Errorf("unable to teardown cluster, it has not been setup")
+	errUnableToStopNotRunningCluster = fmt.Errorf("unable to stop cluster, it is running")
+)
+
+type idToNodeMap map[string]node.ServiceNode
+
+func (im idToNodeMap) values() []node.ServiceNode {
+	returnNodes := make([]node.ServiceNode, 0, len(im))
+	for _, node := range im {
+		returnNodes = append(returnNodes, node)
+	}
+	return returnNodes
+}
+
+type svcCluster struct {
+	sync.RWMutex
+
+	logger       xlog.Logger
+	opts         Options
+	knownNodes   node.ServiceNodes
+	usedNodes    idToNodeMap
+	spares       []node.ServiceNode
+	sparesByID   map[string]node.ServiceNode
+	placementSvc services.PlacementService
+	placement    services.ServicePlacement
+	status       Status
+	lastErr      error
+}
+
+// New returns a new cluster backed by provided service nodes
+func New(
+	nodes node.ServiceNodes,
+	opts Options,
+) (Cluster, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	cluster := &svcCluster{
+		logger:       opts.InstrumentOptions().Logger(),
+		opts:         opts,
+		knownNodes:   nodes,
+		usedNodes:    make(idToNodeMap, len(nodes)),
+		spares:       make([]node.ServiceNode, 0, len(nodes)),
+		sparesByID:   make(map[string]node.ServiceNode, len(nodes)),
+		placementSvc: opts.PlacementService(),
+		status:       ClusterStatusUninitialized,
+	}
+	cluster.addSparesWithLock(nodes)
+
+	return cluster, nil
+}
+
+func (c *svcCluster) addSparesWithLock(spares []node.ServiceNode) {
+	for _, spare := range spares {
+		c.spares = append(c.spares, spare)
+		c.sparesByID[spare.ID()] = spare
+	}
+}
+
+func nodeSliceWithoutID(originalSlice node.ServiceNodes, removeID string) node.ServiceNodes {
+	newSlice := make(node.ServiceNodes, 0, len(originalSlice))
+	for _, elem := range originalSlice {
+		if elem.ID() != removeID {
+			newSlice = append(newSlice, elem)
+		}
+	}
+	return newSlice
+}
+
+// TODO(prateek): reset initial placement after teardown
```

Comment:
The original thought was if we're using the dtest cluster for any other purpose, e.g. (test). It'd be nice to not trample on them. Though even then, I could use a separate env in m3kv for the test and avoid this issue all together. 

Will do that and get rid of this comment.